doc/dev/osd_internals/stale_read.rst

```diff
@@ -0,0 +1,155 @@
+
+On read
+-------
+
+* readable_until = MIN(last_acked_ping for all peers) + read_interval
+* defer read if now >= readable_until.
+
+When sending pg_notify_t
+------------------------
+
+* include last_reply for primary + read_interval - now (this is an
+  upper bound on readable time remaining)
+
+During peering
+--------------
+
+* note peer_readable_until value for all notifies (now + readable time
+  remaining)
+* at activate, for all prior interval members that are in
+  PriorSet.down, add to map osd -> MAX(maybe_rw interval end epoch)
+  (prior_readers)
+* if the resulting map is non-empty, delay activate until
+  MAX(readable_until) for each time whose consumed_epoch is unknown or
+  <= the prior_readers epoch
+
+
+OSD incarnations
+======
+
+Generally speaking, if an OSD is down, the goal is to know that they
+know or are in fact dead.
+
+If an OSD is wrongly marked down, we want to share with our peers the
+consumed_epoch so that we know the PGs got the map and will no longer
+service reads.
+
+If an OSD restarts, we know the old instance must be dead because of
+the exclusive locking on the actual osd data store.
+
+
+Notes
+=====
+
+The good news in all of this is that in the most common case, we are
+marking a peer OSD down because it is failing to respond to
+heartbeats.  By the time we do mark it down, the peers's readable
+intervals will have expired and there will be no further delay in
+peering.
+
+The place where it generally *will* cause an additional delay is when
+we manually mark an OSD down (ceph osd down NNN).  In that case, we
+need to make the downed OSD send a few additional messages to its old
+peers letting them know that it knows about it.  Likewise, the old
+peers should send messages to the downed OSD telling it the news.
+
+ XXX: specifically detail the ping exchanges
+
+
+TODO
+====
+
+#. add heartbeat infrastructure so that we can calculate the
+   readable_until value for an active pg
+#. suspend ops if we have passed the readable_until time
+#. resume ops if we get the relevant heartbeats
+#. share readable_until values along with notify
+#. inherit previous intervals' readable_until on pg activation
+
+#. stress test tool
+
+
+
+
+    P     R
+    |     |
+p1a o     |     first exchange only...
+    |\    |
+    | \   |
+    |  \  |
+    |   \ |
+    |    \|
+    |     o r1 -> r: ru    = -
+    |    /|          ru_ub = r1 + interval
+    |   / |
+    |  /  |
+    | /   |
+    |/    |
+p1b o     |  -> p: ru=p1a + interval, ru_ub=ru
+    |     |
+    |     |
+    |     |
+p2a o     |   for subsequent exchanges, use a mono_clock delta from prior rounds
+    |\    |
+    | \   |
+    |  \  |
+    |   \ |
+    |    \|
+    |     o r2 -> r: ru=r2 + delta_ub + interval
+    |    /|          ru_ub=p2a + delta_ub + interval
+    |   / |
+    |  /  |
+    | /   |
+    |/    |
+p2b o     |  -> p: ru    = p2a + interval
+    |     |        ru_ub = last_ping_sent + interval
```

Comment:
This ru_ub actually gets set (implicitly) when the ping is sent, right?  sent_ping will update last_sent immediately making it available to recalc_readable_until.